### ISSUE TITLE
GitLab Reporter does not report correctly the lines of the errors

### ISSUE THREAD

**Issue by 0x346e3730** [PHP-CS-Fixer/PHP-CS-Fixer#7171]
### Description

GitLab reporter does not report correctly the lines of the errors

### Runtime version

PHP CS Fixer 3.22.0 Chips & Pizza by Fabien Potencier and Dariusz Ruminski.
PHP runtime: 8.1.21

### Used command

```sh
IFS='
'
CHANGED_FILES=$(git diff --name-only --diff-filter=ACMRTUXB "${CI_MERGE_REQUEST_DIFF_BASE_SHA}..HEAD")
EXTRA_ARGS=$(printf -- '--path-mode=intersection\n--\n%s' "${CHANGED_FILES}")
./tools/php-cs-fixer/vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.dist.php -v --dry-run --using-cache=no --format=gitlab ${EXTRA_ARGS} > php-cs-fixer.json
```

### Configuration file

```php
<?php

$finder = PhpCsFixer\Finder::create()
    ->in(__DIR__.'/src/')
    ->in(__DIR__.'/tests/');

$config = new PhpCsFixer\Config();
return $config->setRules([
    '@PSR12' => true,
    '@Symfony' => true,
])
    ->setFinder($finder)
;
```

### Code snippet that reproduces the problem

```php
pUbLiC fUnCtIoN TeStCsFiXeR():void { /* ça devrait trigger de ZIN-ZIN */ $i=0;$i++; }
```

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/28731165/62a5c546-269e-4d4b-a1ac-aa7f964f4e01)

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/28731165/51f15d76-86f4-40a1-a1ee-0d5704fb00b2)

**Comment by Wirone** [comment]
CC: @ostrolucky

**Comment by ostrolucky** [comment]
You have to use the --diff CLI arg for this to work

**Comment by 0x346e3730** [comment]
Indeed, work better this way, thank you and sorry for the wrong issue !

Would it be worth to add this somewhere in the [Using PHP CS Fixer on CI](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/usage.rst#using-php-cs-fixer-on-ci) part of the documentation ? I'm not sure triggering an error in the code if `--format=gitlab` is used and not `--diff` would be nice as I guess it's still a valid usecase ?

**Comment by ostrolucky** [comment]
Makes sense, but there are probably lot of users that use gitlab reporter without --diff flag. Support for correct line numbers was added only very recently, until now it was always 0 no matter what. Raising an error would be a BC break for them. So not sure how to approach it. Good start would be to add it in docs, but I'm also not sure to which part.

**Comment by 0x346e3730** [comment]
However it does not report the correct line number, I have a small offset :

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/28731165/98585a3f-b584-4c27-9417-8a7f3c6aefe6)

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/28731165/c85970bc-9169-4926-8c65-b366ad389377)

I'm doing a PR as a first draft for the documentation addition, we'll be able to discuss there is it's the best place or not

**Comment by ostrolucky** [comment]
Feel free to raise an issue for that one, or create a MR (not sure when will I get around to it). It could very well also be a bug in `sebastian/diff` library that is being used here for this, or the way it's being used.

**Comment by 0x346e3730** [comment]
Here is the first draft for the documentation note : https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7172
Will take a deeper look at the line mismatch when I'll have some time

**Comment by SpacePossum** [comment]
It is impossible, by design, to have the line number per rule. 
For example, rule 1 moves all public and private methods around in the file, than next rule change the casing of the words public and private, what line numbers should be used for what rule?

So you can get line numbers per file (as these we added line numbers to the diff package), but you cannot get these per rule and file.

Many tried anyways, you can find a lot of closed discussions and PRs in this repo about this topic, but you cannot get reliable line numbers per rule. Than again, if the CI trips on CS you can see the diff (and even apply the diff)

**Comment by 0x346e3730** [comment]
When running as `--dry-run` there is no actual fixing so even though there can be a lot of different rules fixes the very first detection on the source file has the actual correct line you would want to output => this should be the one reported. That's theoretically, I have no idea how the code works under the hood yet

**Comment by Wirone** [comment]
Each configured rule is applied to the file in specific order (fixers' priorities), so as it was already pointed out, there's no way to determine line for each rule. Imagine you have line 10 with some code, and then:

- one fixer can move it to different line (like `ordered_class_elements`)
- other fixer can totally change it to different code
- yet another fixer can remove superfluous phpDoc, so everything is shifted up because there's less lines in the file

And when it comes to dry run, lines where rule was triggered may not exist yet, because it would be added at some point during process 😅. Closest result would be if each fixer was run separately on same input file, but in this case some violations wouldn't be even reported, because some fixers would be triggered only because the other fixers were applied before (priorities!).

Even if we applied fixes for each fixer in separate run, and would make dry run before every fix, then it would require some complex logic for tracking line changes, so every fixer's affected lines could be calculated and reported on original lines. Looks terribly complicated for only one feature which's usage is actually unknown 🤷‍♂️.

**Comment by ostrolucky** [comment]
I would say it would make more sense to aggregate all the applied rule names per file and print it as 1 violation in GL reporter.

**Comment by 0x346e3730** [comment]
> I would say it would make more sense to aggregate all the applied rule names per file and print it as 1 violation in GL reporter.

big +1 as this is also what PHPStorm does when reporting PHP CS Fixer errors :
![Screenshot 2023-07-24 at 17 03 28](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/28731165/b142b4d7-ef6d-4343-835a-718f1e9cfd79)

**Comment by JakeQZ** [comment]
> Each configured rule is applied to the file in specific order (fixers' priorities), so as it was already pointed out, there's no way to determine line for each rule. Imagine...

So, IIUC, to get the OP desired behaviour, you'd have to run each rule as a standalone, then report.  Then run the next rule on the original file (as opposed to the modified one), etc.

**Comment by Wirone** [comment]
@JakeQZ no, please read the whole comment. Running each rule separately would not give the same result, as some rules might depend on other rules' outcome.

**Comment by keradus** [comment]
I'm thinking about this every few weeks.

Perhaps, `php-cs-fixer check` only run rules separately and tho be able to give the accurate line numbers, why it is OK to not report issues that would be visible if we would actually apply rule after rule.

To fix all issues, I expect single run of `fix`.
To detect issues, I expect run of `check` (locally or CI), and then have it fixed by... `fix` - so anyway, next `check` wouldn't find anything.

**Comment by Wirone** [comment]
@keradus as it was pointed out several times already, running `check` with each single rule separately won't give you the same result as running `check` with all the rules at once, because some rules may be triggered only because previous rules did something (priorities, triggering tokens).

**Comment by JakeQZ** [comment]
> each single rule separately won't give you the same result as running `check` with all the rules at once, because some rules may be triggered only because previous rules did something (priorities, triggering tokens).

I understand.  However, when used as part of continuous integration, there are unlikely to be multiple issues to be fixed.

That said, I find the `--diff` option with `--dry-run` to be adequate for identifying the location of the problem.

But without it, we are left somewhat in the dark.  An approximate line number could still be useful, even if potentially inaccurate due to knock-on effects.

> 1\) tests\Value\SizeTest.php (function_declaration, php_unit_fqcn_annotation)
> Found 1 of 52 files that can be fixed

The reportage with `(function_declaration, php_unit_fqcn_annotation)` is vague and gives no clue where the problem actually is.  (It doesn't even relate to a function declaration, but a DocBlock for a class.)

**Comment by connorhu** [comment]
After some debugging, I managed to create a perfect solution with the example code. Now it can guess character positions.
This testing is just to see if the algorithm works without modifying the php-cs-fixer and if it puts the markup in the right place.
I will look for more complex example code to test the idea further.
```
line:  6 char:  0 rule: no_superfluous_phpdoc_tags
line: 10 char: 45 rule: void_return
line:  5 char:  0 rule: no_empty_phpdoc
line:  1 char:  6 rule: declare_strict_types
line:  1 char:  0 rule: blank_line_after_opening_tag
line: 12 char: 13 rule: yoda_style
line: 15 char:  9 rule: no_mixed_echo_print
line:  3 char:  7 rule: psr_autoloading
line: 13 char:  0 rule: no_extra_blank_lines
line:  3 char:  0 rule: header_comment
```

**Comment by zyberspace** [comment]
*Activity*

(It's still an issue)

**Comment by JakeQZ** [comment]
> _Activity_
> 
> (It's still an issue)

The bot should check for activity in a linked PR - #7865.

I think the solution based on overall `diff` may be cleanest.  It should be able to provide output of "between line A and B" even if not a precise line numner - unless the are fixers that actually reorder the code - e.g. putting all `public` methods before `protected`.

An alternative might be recording the original line number with each line, and preserving those through all fixer changes, with possibly multiple line numbers applying to some resultant lines, and possibly some resultant lines sharing the same original line numbers.  But that would probably still require an major architecural change to deal with function/method reordering.

WinMerge's moved code block feature works well, and it is open source.  That might be of use...

FOAD, I'm not familiar with the souce code/archtecture.  My knowledge is based on usage and what I've read here and in the docs.  But I can imagine that method reordering is one possible fixer that would pose problems.

**Comment by abodnar** [comment]
This is still an issue with the gitlab formatter.

Example:

```
<?php

namespace Database\Seeders;

use App\Models\Client;
use Illuminate\Database\Seeder;

class ClientSeeder extends Seeder
{
    /**
     * Run the database seeds.
     *
     * @return void
     */
    public function run()
    {
        //Client with ridiculously long name. Constant name used for search.
        $client = Client::factory()
            ->create([
                'name' => 'Test Client',
            ]);
    }
}
```

When the json output is returned this shows as line 14 being the problem, when in fact it should be line 17.

```
[
    {
        "check_name": "PHP-CS-Fixer.single_line_comment_spacing",
        "description": "PHP-CS-Fixer.single_line_comment_spacing by PHP CS Fixer 3.68.5 Persian Successor by Fabien Potencier, Dariusz Ruminski and contributors.",
        "categories": [
            "Style"
        ],
        "fingerprint": "d4720bb9a19231cdfbf19b064cb70099",
        "severity": "minor",
        "location": {
            "path": "/var/www/html/database/seeders/ClientSeeder.php",
            "lines": {
                "begin": 14,
                "end": 7
            }
        }
    }
]
```

**Comment by tontonsb** [comment]
Hey all, the current offset in the Gitlab report is caused by the diff "context" — the differ includes 3 lines of context before and after the changed part. That's why `location.lines.begin` is 3 lines before the changed one and why `location.lines.end` usually indicates a block of 7 lines. Unfourtunately Gitlab links to the first line.

I'm not sure if the `end` is correct as 7 or it should've been equal to `begin+7`, but I think the context lines should just be removed so Gitlab can link to the correct lines and display context if it decides to.

I've sketched some work on this in #8464
